flocker/apiclient/_client.py

```diff
@@ -144,4 +150,127 @@ def synchronize_state(self):
+                                              cert_path, key_path)
+        self._base_url = b"https://{}:{}/v1".format(host, port)
+
+    def _request(self, method, path, body, success_code, error_codes={}):
+        """
+        Send a HTTP request to the Flocker API, return decoded JSON body.
+
+        :param bytes method: HTTP method, e.g. PUT.
+        :param bytes path: Path to add to base URL.
+        :param body: If not ``None``, JSON encode this and send as the
+            body of the request.
+        :param int success_code: Expected success response code.
+        :param error_codes: Mapping from HTTP response code to exception to be
+            raised if it is present.
+
+        :return: ``Deferred`` firing with decoded JSON.
+        """
+        def error(body, code):
+            if code in error_codes:
+                raise error_codes[code](body)
+            raise ResponseError(code, body)
+
+        def got_result(result):
+            if result.code == success_code:
+                return json_content(result)
+            else:
+                d = content(result)
+                d.addCallback(error, result.code)
+                return d
+
+        headers = None
+        data = None
+        if body is not None:
+            headers = {b"content-type": b"application/json"}
+            data = dumps(body)
+        request = self._treq.request(
+            method, self._base_url + path,
+            data=data, headers=headers,
+            # Keep tests from having dirty reactor problems:
+            persistent=False
+        )
+        request.addCallback(got_result)
+        return request
+
+    def _parse_configuration_dataset(self, dataset_dict):
+        """
+        Convert a dictionary decoded from JSON with a dataset's configuration.
+
+        :param dataset_dict: Dictionary describing a dataset.
+        :return: ``Dataset`` instance.
+        """
+        return Dataset(primary=UUID(dataset_dict[u"primary"]),
+                       maximum_size=dataset_dict[u"maximum_size"],
+                       dataset_id=UUID(dataset_dict[u"dataset_id"]),
+                       metadata=dataset_dict[u"metadata"])
+
+    def create_dataset(self, primary, maximum_size, dataset_id=None,
+                       metadata=pmap()):
+        dataset = {u"primary": unicode(primary),
+                   u"maximum_size": maximum_size,
+                   u"metadata": dict(metadata)}
+        if dataset_id is not None:
+            dataset[u"dataset_id"] = unicode(dataset_id)
+
+        request = self._request(b"POST", b"/configuration/datasets",
+                                dataset, CREATED,
+                                {CONFLICT: DatasetAlreadyExists})
+        request.addCallback(self._parse_configuration_dataset)
+        return request
+
+    def move_dataset(self, primary, dataset_id):
+        request = self._request(
+            b"POST", b"/configuration/datasets/{}".format(dataset_id),
```

Comment:
:question: Maybe don't use `bytes.format` or maybe use unicode here and elsewhere in this branch when specifying paths for requests.